src/compile/data/optimize.ts

```diff
@@ -140,31 +148,51 @@ export function mergeParse(node: DataFlowNode) {
       }
     }
   }
-  node.children.forEach(mergeParse);
+  return node.children.map(mergeParse).some(x => x === true) || flag;
 }
 
 /**
  * Optimizes the dataflow of the passed in data component.
  */
-export function optimizeDataflow(dataComponent: DataComponent) {
-  let roots: SourceNode[] = vals(dataComponent.sources);
-
-  roots.forEach(removeUnnecessaryNodes);
-
-  // remove source nodes that don't have any children because they also don't have output nodes
-  roots = roots.filter(r => r.numChildren() > 0);
-  getLeaves(roots).forEach(optimizers.iterateFromLeaves(optimizers.removeUnusedSubtrees));
-  roots = roots.filter(r => r.numChildren() > 0);
-
-  getLeaves(roots).forEach(optimizers.iterateFromLeaves(optimizers.moveParseUp));
-  getLeaves(roots).forEach(optimizers.removeDuplicateTimeUnits);
-
-  roots.forEach(moveFacetDown);
-  roots.forEach(mergeParse);
-
-  keys(dataComponent.sources).forEach(s => {
-    if (dataComponent.sources[s].numChildren() === 0) {
-      delete dataComponent.sources[s];
+export function optimizeDataflow(data: DataComponent) {
+  function helper(dataComponent: DataComponent) {
```

Comment:
The method name helper is a bit confusing.  Basically this is a method that run one pass.  Give it a better name?

Reply:
Also, look likes it's unnecessary to nest this method inside `optimizeDataflow`.  So don't nest it then.